src/game/player_callbacks_1.c

```diff
@@ -302,4 +325,491 @@ void PlayerCB_8025854(Player *player)
+    if (IsBossStage(gCurrentLevel)) {
+        if ((player->moveState & MOVESTATE_IN_AIR)) {
+            sub_8025F84(player);
+            return;
+        }
+    }
+
+    if ((player->moveState & (MOVESTATE_800 | MOVESTATE_8 | MOVESTATE_IN_AIR))
+        == MOVESTATE_800) {
+        sub_802A360(player);
+    } else {
+        player->unk90->s.unk10 &= ~SPRITE_FLAG_MASK_14;
+        player->unk64 = 4;
+
+        sub_8023B5C(player, 9);
+
+        player->unk16 = 6;
+        player->unk17 = 9;
+
+        player->moveState |= MOVESTATE_4;
+        player->unk99 = 0;
+        gPlayer.callback = PlayerCB_8025AB8;
+        gPlayer.callback(player);
+    }
+}
+
+void PlayerCB_8025AB8(Player *player)
+{
+    if ((player->moveState & (MOVESTATE_8000000 | MOVESTATE_8 | MOVESTATE_IN_AIR))
+        == MOVESTATE_8000000) {
+        PlayerCB_80273D0(player);
+    } else if ((player->moveState & (MOVESTATE_800 | MOVESTATE_8 | MOVESTATE_IN_AIR))
+               == MOVESTATE_800) {
+        sub_802A360(player);
+    } else {
+        if (player->unk99) {
+            player->unk99--;
+        } else if (!sub_8029E6C(player)) {
+            if (player->unk2A == 0) {
+                u16 dpadSideways = (player->unk5C & (DPAD_LEFT | DPAD_RIGHT));
+                if (dpadSideways != DPAD_RIGHT) {
+                    if (dpadSideways == DPAD_LEFT) {
+                        s32 val = player->speedGroundX;
+                        if (val <= 0) {
+                            player->moveState |= MOVESTATE_FACING_LEFT;
+                        } else if ((val - Q_24_8(0.09375)) < 0) {
+                            s32 deceleration = -Q_24_8(0.375);
+                            player->speedGroundX = deceleration;
+                        } else {
+                            player->speedGroundX = (val - Q_24_8(0.09375));
+                        }
+                    }
+                } else {
+                    s32 val = player->speedGroundX;
+                    if (val >= 0) {
+                        player->moveState &= ~MOVESTATE_FACING_LEFT;
+                    } else if ((val + Q_24_8(0.09375)) > 0) {
+                        player->speedGroundX = Q_24_8(0.375);
+                    } else {
+                        player->speedGroundX = (val + Q_24_8(0.09375));
+                    }
+                }
+            }
+        } else {
+            return;
+        }
+
+        if (player->speedGroundX > 0) {
+            player->unk50 = 8;
+        } else if (player->speedGroundX < 0) {
+            player->unk50 = -8;
+        } else {
+            player->unk50 = 0;
+        }
+
+        player->speedGroundX -= player->unk50;
+
+        if ((player->speedGroundX > -Q_24_8(0.5))
+            && (player->speedGroundX < Q_24_8(0.5))) {
+            player->unk50 = 0;
+            player->speedGroundX = 0;
+        }
+
+        if (player->speedGroundX == 0) {
+            gPlayer.callback = PlayerCB_8025318;
+            gPlayer.callback(player);
```

Comment:
Does this match with `PlayerCB_8025318(player)`? just wondering

Reply:
Huh, it actually does.
I only tried player->callback(player) before, which didn't work.

Guess it makes more sense to write it this way?